### ISSUE TITLE
Mac OS support

### ISSUE THREAD

**Issue by brunopacheco1** [OlegKunitsyn/gnucobol-debug#18]
cobc (GnuCOBOL) 2.2.0
GNU gdb (GDB) 9.1

Running this [sample](https://github.com/brunopacheco1/cobol-hello-world/blob/master/petstore.cob) on a MacOS, there's a strange bug happing when running the code(? not sure, just a guess). Please refer to the output attached.

This a log extracted from the log: 
stdin: 9-break-insert -f "/Users/bpacheco/dev/personal_git/cobol-hello-world/petstore.c:106"
stdout: 7^error,msg="malformed linespec error: unexpected colon"

[debbuger_output_macos.txt](https://github.com/OlegKunitsyn/gnucobol-debug/files/4551843/debbuger_output_macos.txt)

**Comment by gfmoore** [comment]
Yep, that's the error I'm getting (see the issue #16) . I know where it is (I think) but I need to debug the debugger. I've downloaded the code from github, and I've set of TypeScript and I'm now trying to get it to work.

I'm getting an error though when I set a breakpoint on mi2.ts where though I set a breakpoint when I debug the debugger it launches a vs code instance, I run my cobol test script, but the breakpoints on mi2.ts disappear and I get yet another error "breakpoint set, but no yet bound".

Sorry I've gone off topic, but it's on line 460 of mi2.ts where a colon gets added and I want to know what's in the variable that follows.

Sorry I'm hacking around here.

**Comment by brunopacheco1** [comment]
@OlegKunitsyn what do you think of having a channel (Slack?) for this kind of discussion?

**Comment by brunopacheco1** [comment]
@gfmoore It looks like we have faced similar issues by luck, but they are not related to the Mac OS support. I have updated my first comment to include the correct log output. Please, keep the discussions about the bug you were facing on issue  #16 .

**Comment by OlegKunitsyn** [comment]
> @OlegKunitsyn what do you think of having a channel (Slack?) for this kind of discussion?

We may try. The method does not affect the quantum of free time.

**Comment by brunopacheco1** [comment]
Sure it doesn't, I just thought to avoid polluting the issue comments with not related topics, but in the end it doesn't change anything.

**Comment by OlegKunitsyn** [comment]
Fine to me. Just open new one and keep posting. What about support of COBOL archive files .car and unit-testing?

**Comment by GitMensch** [comment]
> COBOL archive files .car

???

> What about [...] unit-testing?

Everything that concerns linting/mapping would be useful to be unit tested (and depending on the current code structure may would have to be moved out of other code, if those are "mixed"), for everything else one would need to either hook into vscode execution (is there a way to do this "from outside") or use a ui testing framework (which I'd say is likely too much for now).

**Comment by brunopacheco1** [comment]
When debugging on MacOS, I received this message: "No symbol table is loaded.  Use the "file" command.". Looking out there on Internet, it looks the program should have been compiled using the arg "-ggdb". 

I have set COB_CFLAGS and COB_LDFLAGS to add this arg, but I've got the same error.

`{
            "name": "COBOL debugger",
            "type": "gdb",
            "request": "launch",
            "target": "${file}",
            "targetargs": [],
            "cwd": "${workspaceRoot}",
            "env": { "COB_CFLAGS" :"-ggdb" },
            "gdbpath": "gdb",
            "cobcpath": "cobc",
            "cobcargs": ["-x"],
            "group": [],
            "verbose": true
        }`

@GitMensch Do you know what that means?

**Comment by brunopacheco1** [comment]
Paying a bit more attention on the log, I could see this line.

`GDB -> App: {"outOfBandRecord":[{"isStream":true,"type":"log","content":"warning: `/var/folders/47/cwm092rj3ms52r4dv9j66n5h0000gn/T//cob21532_0.o': can't open to read symbols: No such file or directory.\n"}]}`

Does the GDB need to have object files? Isn't it attaching directly to the running process?

**Comment by brunopacheco1** [comment]
`cclang: warning: /var/folders/47/cwm092rj3ms52r4dv9j66n5h0000gn/T//cob23811_0.o: 'linker' input unused [-Wunused-command-line-argument]
clang: warning
Could not start debugger process
: -lcob: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: -lm: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: -lgmp: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: -lncurses: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: -ldb-18.1: 'linker' input unused [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-ggdb' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/opt/gmp/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/opt/berkeley-db/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/usr/local/Cellar/gnu-cobol/2.2_1/lib' [-Wunused-command-line-argument]`

**Comment by GitMensch** [comment]
I don't know much about the clang environment. You may or may not need additional flags to compile "GDB compatible". In any case: try with the gdb command line first before checking this extension.
`COB_CFLAGS` would be wrong as this overrides the built-in default specs in the cobc binary. Better pass `-A "-whateverflagyouwanttoadd"` to `cobc` instead.
And it looks like you use an older version (the current 3.1-dev should not try to link gmp, ncurses or bdb; to also remove the libmath reference you'd currently have to do a manual step in building GnuCOBOL.

**Comment by brunopacheco1** [comment]
Thanks for the suggestions, I'll keep checking. But yes, on Mac I was quite lazy to get the latest GC and relied on BREW for that.

**Comment by GitMensch** [comment]
> I was quite lazy to get the latest GC and relied on BREW for that.

In this case I suggest you post fix these:
`cobc --info` -> note down `COB_LDFLAGS` (will likely include the name and path to unnecessary libraries.) then export `COB_LDFLAGS` without these (in a central place).

**Comment by GitMensch** [comment]
@brunopacheco1 if I understand correctly this issue is more about clang support (in this case you could also reproduce the issue on GNU/Linux and possibly wsl2 when self-building GnuCOBOL and specifying CC=clang).
If this is correct it would be good to adjust the issue title.

**Comment by cribetti** [comment]
I am having the same symptoms on mac as @brunopacheco1. from his attached output file in the first comment through the attempts to resolve the issue, it looks identical to what i am seeing. i am unable to resolve. i can run without the debugger but once the debugger is included in the run all the above symptoms appear. let me know how i may assist in resolving this issue.

**Comment by GitMensch** [comment]
@cribetti I guess your issue is too about clang support, not about MacOS in general. Correct? In this case I still argue that @brunopacheco1 or @OlegKunitsyn may should change the title and that someone on MacOS retries with a GCC setup.

**Comment by brunopacheco1** [comment]
> @cribetti I guess your issue is too about clang support, not about MacOS in general. Correct? In this case I still argue that @brunopacheco1 or @OlegKunitsyn may should change the title and that someone on MacOS retries with a GCC setup.

Sorry @GitMensch I didn't have time to tackle this issue yet. @cribetti if you could double-check what Simon has mentioned, that would be quite helpful already.

**Comment by cribetti** [comment]
@GitMensch i do not know where the source of the problem is. i've been looking for a solution on the internet and came across this thread and found someone who has what seems to be the same issue. i have a question: what does this mean?

warning: `/var/folders/sv/0361lt0n5r94zdyqwz19d_580000gn/T//cob6335_0.o': can't open to read symbols: No such file or directory.
BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34

my error log has this as my main focus. I've looked in the folder and there's a tons of unrelated files, but no .o file.

@brunopacheco1 you can get rid of the codesign errors by following: https://sourceware.org/gdb/wiki/PermissionsDarwin

i can help with things. what would simon like me to double check? please be very specific so i can verify accurately.

I installed cobc with ```brew install gnu-cobol```. i did not ```make```

here is my environment:

```
gnu-cobol/2.2_1 % cobc -info                                                                                                                                            ] 7:32 PM
cobc (GnuCOBOL) 2.2.0
Copyright (C) 2017 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Keisuke Nishida, Roger While, Ron Norman, Simon Sobisch, Edward Hart
Built     Apr 29 2020 00:06:51
Packaged  Sep 06 2017 18:48:43 UTC
C version "4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.32.59)"

build information
build environment        : x86_64-apple-darwin19.4.0
CC                       : clang
C version                : "4.2.1 Compatible Apple LLVM 11.0.3
                           (clang-1103.0.32.5
CPPFLAGS                 : -I/usr/local/opt/gmp/include
                           -I/usr/local/opt/berkeley-db/include
CFLAGS                   : -O2 -pipe
LD                       : /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
LDFLAGS                  : -L/usr/local/opt/gmp/lib
                           -L/usr/local/opt/berkeley-db/lib

GnuCOBOL information
COB_CC                   : clang
COB_CFLAGS               : -I/usr/local/Cellar/gnu-cobol/2.2_1/include
                           -Wno-unused -fsigned-char -Wno-pointer-sign
                           -Qunused-arguments
                           -I/usr/local/opt/gmp/include
                           -I/usr/local/opt/berkeley-db/include -pipe
COB_LDFLAGS              : -L/usr/local/opt/gmp/lib
                           -L/usr/local/opt/berkeley-db/lib
COB_LIBS                 : -L/usr/local/opt/gmp/lib
                           -L/usr/local/opt/berkeley-db/lib
                           -L/usr/local/Cellar/gnu-cobol/2.2_1/lib -lcob
                           -lm -lgmp -lncurses -ldb-18.1
COB_CONFIG_DIR           : /usr/local/Cellar/gnu-cobol/2.2_1/share/gnucobol/config
COB_COPY_DIR             : /usr/local/Cellar/gnu-cobol/2.2_1/share/gnucobol/copy
COB_MSG_FORMAT           : GCC
COB_OBJECT_EXT           : o
COB_MODULE_EXT           : dylib
COB_EXE_EXT              : 
64bit-mode               : yes
BINARY-C-LONG            : 8 bytes
extended screen I/O      : ncurses
variable format          : 0
sequential handler       : built-in
ISAM handler             : BDB
mathematical library     : GMP
gnu-cobol/2.2_1 %           
```

**Comment by GitMensch** [comment]
We do see that all environments that are mentioned in this issue share the following;

* Mac OS
* used brew packages (I *think* those are binary packages, please correct me if I'm wrong)
* the binary was built with clang
* this plugin triggers GDB but GDB can't find symbols

From all these points I *guess* the issue is either because gdb can't work as expected with the clang generated object files or there's an issue with the debug configuration.

You *could* do the following to activate the GDB-tuning: use launch.json, add the following two arguements `"-A\"-ggdb\"", "-Q\"-ggdb""` (possibly only one needed) and see if this makes any difference.

You could also *temporarily* add `"-v"` or even `"-vv"` / `"-vvv"` which is known to break the debugging part in any case (#25) but may provide some useful hints .

The ideal check afterwards would be to use `make` to actually build GnuCOBOL from source (for now you'd use a [3.1 snapshot](https://ci.appveyor.com/api/projects/GitMensch/gnucobol-3-x/artifacts/gnucobol-3.1-dev.tar.gz?job=Image:%20Ubuntu1804) and if you see or have any issues doing so report via bug list or discussion boards [see `cobc --help`]) - and using gcc as environment (at least after verifying if the self-created version that uses `clang` does/does not work).

**Comment by brunopacheco1** [comment]
I have tried changing clang for gcc and the error message was the following: `gcc-9: error: unrecognized command line option '-Qunused-arguments'`. I have overridden the gcc arguments to remove this one (which is specific for CLANG) and I have also added "-save-temps" to cobc call.

The error message now is: 
`GDB -> App: {"token":3,"outOfBandRecord":[],"resultRecords":{"resultClass":"done","results":[["source-path","/Users/bpacheco/dev/personal_git/cobol-hello-world:$cdir:$cwd"]]}}
BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34


BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34

BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34`

I have done the following tries:
* brew GC 2.2, CLANG, cobc -save-temps
* locally compiled with CLANG GC 3.1, CLANG, cobc -save-temps
* locally compiled with CLANG GC 3.1, GCC, cobc -save-temps, overriding -Qunused-arguments

I've got exactly the same error message for all cases.

But so far, it looks like the cobc compiles and the debugger extension loads, but GDB is failing. Does it support CLANG? I am not familiar with C/C++ environment, perhaps this is obvious for someone else.

Now I am going to try compiling gnucobol with GCC.

**Comment by brunopacheco1** [comment]
Well, compiling gnucobol with gcc gives me exactly the same error message. Perhaps the issue is related to this dyld instead? What do you think?

**Comment by GitMensch** [comment]
> Well, compiling gnucobol with gcc and running the debugger gives me exactly the same error message.

You did recognize that it is just the gcc-frontend of clang, didn't you? The compiler and linker is exactly the same (at least as far as I know the gcc-fronted mainly handles the options where they differ and adds some additional defines and macros). But I may be wrong here. [BTW: I'd like to have a look at tests/testsuite.log from GnuCOBOL which is the result of `make check` - if you don't mind]

Hm, do you see any errors like the following on GNU/Linux, too?

~~~
BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34
BFD: /usr/lib/dyld(i386:x86-64)
: unknown load command 0x34
BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34
~~~

The debugger *may* have additional issues with the signature error:

~~~
Unable to find Mach task port for process-id 54707: (os/kern) failure (0x5).
(please check gdb is codesigned - see taskgated(8))
~~~

Did you checked `taskated(8)` as gdb hints or tried the approach from @cribetti (https://sourceware.org/gdb/wiki/PermissionsDarwin) already?

**Comment by cribetti** [comment]
> The debugger _may_ have additional issues with the signature error:
> 
> ```
> Unable to find Mach task port for process-id 54707: (os/kern) failure (0x5).
> (please check gdb is codesigned - see taskgated(8))
> ```
> 
> Did you checked `taskated(8)` as gdb hints or tried the approach from @cribetti (https://sourceware.org/gdb/wiki/PermissionsDarwin) already?

this will get resolved by following the steps in https://sourceware.org/gdb/wiki/PermissionsDarwin, especially step 1.3. other internet solutions about codesigning are incomplete and the above message will remain unless the above steps are followed.

the remaining message
```BFD: /usr/lib/dyld(i386:x86-64): unknown load command 0x34```
may provide a direction for resolution?

**Comment by GitMensch** [comment]
> I think we have to bundle the extension with a "redistributable" versions of cobc etc for all these 3 platforms,

I strongly advice against bundling GnuCOBOL with this extension as people want to debug their system programs which uses whatever version of GnuCOBOL is locally installed.
It would be very reasonable to tell people "these are the binary packages we've tested against" and provide an HOWTO debug/handle problems for local environments.

**Comment by GitMensch** [comment]
@brunopacheco1 Do you have the option to recheck with a "real" GCC (not a clang front-end)?

**Comment by OlegKunitsyn** [comment]
> I strongly advice against bundling GnuCOBOL with this extension as people want to debug their system programs which uses whatever version of GnuCOBOL is locally installed.
> It would be very reasonable to tell people "these are the binary packages we've tested against" and provide an HOWTO debug/handle problems for local environments.

People are different. Experienced devs can code in Notepad, but beginners need more love. 

Story at work: until I started bundling lib-sensitive Java apps with redistributable jvm for nix/win/mac I was receiving hundreds or strange and unreproducible errors via telemetry. The users (software engineers and data analysts, btw) were disappointed and left unhappy. For example, JetBrains and Elastic bundle jvms, too.

Downloads/installations rate of the extension is too high. I see no other way to provide reasonable support with available resources. Portability will help write, execute and debug educational helloworld programs immediately. At the same time, gurus still be able to point cobcpath and gdbpath in launch.json to anything. Win-win?

**Comment by brunopacheco1** [comment]
> @brunopacheco1 Do you have the option to recheck with a "real" GCC (not a clang front-end)?

I have to double-check, but gcc-9 present on brew was supposed to be the real GCC.

**Comment by GitMensch** [comment]
Maybe I've missed your update. The last one I've recognized came with a log.txt  which shows that the system was *not* build with gcc:

~~~
CC                       : gcc
C version                : "4.2.1 Compatible Apple LLVM 11.0.3
                           (clang-1103.0.32.2)
~~~

If you used a different one (possibly by specifying something like `CC=gcc-9` when running the configure script) then I've just missed it.

**Comment by brunopacheco1** [comment]
I think I've mentioned that, but you are right, I forgot to share the logs of my tests, sorry about that. Tomorrow I'll update you guys.

**Comment by brunopacheco1** [comment]
This is the log after recompiling GC 3.1 with GCC-9 and signing GDB.

[gcc9-log.txt](https://github.com/OlegKunitsyn/gnucobol-debug/files/4622435/gcc9-log.txt)

Now, the GDB freezes and doesn't reply with threads-info request.

**Comment by brunopacheco1** [comment]
I will try getting the GDB source and compiling everything.

**Comment by brunopacheco1** [comment]
This is the log with GDB and GC compiled locally with GCC-9.

[gcc9-log_compiled_gdb.txt](https://github.com/OlegKunitsyn/gnucobol-debug/files/4622723/gcc9-log_compiled_gdb.txt)

On the first try, there are two `thread-info` requests and no GDB response. On the second try, with locally compiled GDB, the first one has got an answer, but the second no.

I'll keep checking it, but so far my ideas are:
- There's a concurrency issue in the debugger;
- There is a permission issue in the security policy defined for GDB;
- The debugger is hiding another error;

**Comment by GitMensch** [comment]
Thank you for this work, I also learned some thing "on the way" by just reading and doing some additional search. I know also now that there are different load commands in MacOS and that GDB will prompt on any unknown load command. I've also found different older headers, but the 0x34 seems to be quite new (as your Mac OS is too - I assume).
The common suggestions are "ignore, likely everything will just work", "use an Apple distributed version of GDB" (is there any new one - did anyone tried with a apple-distributed one?) and "apple supports lldb - use this" (and I've read that it supports MI, but no clue about MI2).

**Comment by brunopacheco1** [comment]
Yes, I am ignoring the load commands for now, and focusing on GDB and security policies. 

I've got the following errors when watching MacOSTaskPolicy logs:

`2020-05-14 10:08:53.143057+0200 0x1c762    Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:53.228971+0200 0x123d     Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:53.232970+0200 0x1c762    Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:57.463442+0200 0x123d     Activity    0x1049d              580    0    taskgated: (Security) SecTrustEvaluateIfNecessary
2020-05-14 10:08:57.468027+0200 0x123d     Activity    0x1049e              580    0    taskgated: (Security) SecTrustEvaluateIfNecessary
2020-05-14 10:08:57.469592+0200 0x1e752    Default     0x0                  0      0    kernel: (AppleMobileFileIntegrity) macOSTaskPolicy: (gdb) (pid: 3690) not (yet) allowed to go invalid
2020-05-14 10:08:57.470987+0200 0x1c762    Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:57.479629+0200 0x123d     Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:57.487496+0200 0x1c762    Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062
2020-05-14 10:08:57.500443+0200 0x123d     Default     0x0                  580    0    taskgated: (Security) [com.apple.securityd:security_exception] MacOS error: -67062`

Maybe it is still a security issue. When signing GDB, you have to define what permissions you want to give, and most of the blogs suggest to add just "com.apple.security.cs.debugger" in the entitlements list, but perhaps we should add specifics permissions for different requirements.

I will spend a bit more time on that before checking out different GDB's and further on LLDB.

**Comment by brunopacheco1** [comment]
It's impressive how many people have GDB debugging issues on MacOS, recurrently on each new OS version.

And the best comment so far I've was on a Chinese page, the guy says "Stop fighting, use LLDB, you don't want to waste time on such things".

**Comment by GitMensch** [comment]
> It's impressive how many people have GDB debugging issues on MacOS, recurrently on each new OS version.

Yes, what a co-incidence, given that Apple wants to push LLDB (and the complete LLVM stack) as they can integrate it into XCode and other proprietary tools which is (by design!) not possible with GDB...

> the best comment so far I've was on a Chinese page, the guy says "Stop fighting, use LLDB, you don't want to waste time on such things".

Yes, that's a possible way (and likely what Apple want to... strongly encourage everybody to do, which has "likely" nothing to do with the first argument).

Just to make it clear: I don't have anything against LLDB in general (and also suggested to *additional* support it, if it isn't too much work; mostly because it claims it works fine with pdb-files generated by another "popular" proprietary compiler).

And yes - those permission errors are likely good to be solved in any case...

**Comment by OlegKunitsyn** [comment]
Gentlemen, please test containerized GnuCOBOL with 0.22.15 on Mac/Win if you can.

**Comment by brunopacheco1** [comment]
It is working correctly on Mac using docker. I'll doble-check on Windows, but my guess if it works over WSL2, it is going to work also over docker there.

<img width="1680" alt="Screenshot 2020-05-17 at 10 22 05" src="https://user-images.githubusercontent.com/4491850/82139498-72c75000-9828-11ea-8931-1edfd1c86ac5.png">

**Comment by brunopacheco1** [comment]
Sharing my latest tests on Mac using local binaries, I have added more permissions entitlements, but the errors persist. [GDB on Mac OS](https://timnash.co.uk/getting-gdb-to-semi-reliably-work-on-mojave-macos/)

I would say it doesn't worth spending more time on making it work on MacOS using GDB. Unfortunately, GDB is buggy on Mac and even it is fixed now, by the tons of complaints on the Internet, they will break GDB after a new MacOS release.

The GC's binary on BREW is already compiled using CLANG, so perhaps we LLDB could be supported by this extension, what do you think?

**Comment by brunopacheco1** [comment]
It is also working on docker/Windows as expected.

![image](https://user-images.githubusercontent.com/4491850/82143243-baf36c00-9842-11ea-9a2a-a4e8a9dae3e3.png)

**Comment by brunopacheco1** [comment]
Perhaps you could suggest adding a Task on VSCode, something like:

`{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "run gnucobol container",
            "type": "shell",
            "command": "docker run -d -i --name gnucobol -w ${workspaceRoot} -v ${workspaceRoot}:${workspaceRoot} olegkunitsyn/gnucobol:3.1-dev"
        },
        {
            "label": "stop gnucobol container",
            "type": "shell",
            "command": "docker stop gnucobol"
        }
    ]
}`

**Comment by brunopacheco1** [comment]
BTW, thanks a lot, it's quite handy having everything in one single place.

**Comment by GitMensch** [comment]
> perhaps LLDB could be supported by this extension, what do you think?

As written before it does implement MI, but possibly not MI2 (I don't know); if this plugin can be made to work with both GDB and LMDB (please don't drop the first!) then it would be good to do so.
@brunopacheco1 Can you please create a new issue for this? Would you like to start working on this after the numeric field display/accept currently part of #4 is finished (if @OlegKunitsyn say he'd like to include LMDB if possible, of course)?

**Comment by OlegKunitsyn** [comment]
> Perhaps you could suggest adding a Task on VSCode, something like:
Committed

**Comment by OlegKunitsyn** [comment]
> The GC's binary on BREW is already compiled using CLANG, so perhaps LLDB could be supported by this extension, what do you think?

AFAIK LLDB uses similar set of commands. We may try to abstract them, add a 'lldb' config and  implement LldbDebugSession.

**Comment by OlegKunitsyn** [comment]
lldb branch is a hack that demonstrates `lldb-mi` in action. Breakpoints and coverage looks OK, but not the variables. The pane is empty and here's the reason:

```
# GDB/MI2
13-stack-list-variables --thread 1 --frame 0 --simple-values
13^done,variables=[
{name="entry",arg="1",type="const int",value="0"},
{name="initialized",type="unsigned int",value="1"},
{name="module",type="cob_module *",value="0x55555556be70"},
{name="cob_glob_ptr",type="cob_global *",value="0x55555555d6b0"},
{name="cob_procedure_params",type="cob_field *[1]"},
{name="frame_overflow",type="struct cob_frame *",value="0x7fffffffdc80"},
{name="frame_ptr",type="struct cob_frame *",value="0x7fffffffcca0"},
{name="frame_stack",type="struct cob_frame [255]"},
{name="b_2",type="int",value="0"},
{name="b_6",type="unsigned char [5]"},
{name="f_6",type="cob_field"}
]

# LLDB-MI-9
11-stack-list-variables --thread 1 --frame 0 --simple-values
11^done,variables=[
{arg="1",name="entry",type="const int",value="0"},
{name="cob_glob_ptr",type="cob_global *"},
{name="cob_procedure_params",type="cob_field *[1]"},
{name="frame_overflow",type="cob_frame *"},
{name="frame_ptr",type="cob_frame *"},
{name="frame_stack",type="cob_frame [255]"}
]

```

`-stack-list-variables` does not return the variables. I've tested on LLDB-9, LLDB-8, LLDB-7. LLDB-10 does not include LLDB-MI at all. Same on Alpine image. Please clone, backtrace and fix if you can.

**Comment by GitMensch** [comment]
> LLDB-10 does not include LLDB-MI at all.

It was moved out of the main LLDB source to a separate project https://github.com/lldb-tools/lldb-mi, see http://lists.llvm.org/pipermail/lldb-dev/2019-August/015357.html for the announcement on this,

It should be possible to use a snapshot of this repo to add it to LLDB-10 and _maybe_ even possible that way to "update" the MI driver for older LLDB versions.
